### ISSUE TITLE
DOC: Guidance on available quadrature methods

### ISSUE THREAD

**Issue by vnmabus** [scipy/scipy#19368]
### Issue with current documentation:

The current documentation lists all available quadrature methods (given function object) in SciPy. However, it does not offer any guidance on which one to choose. I assume that most people just use `quad` because of this.

### Idea or request for content:

Please add a table in the documentation that compares available quadratures so that we can easily check:
- If they support vectorized functions.
- If they support vector-valued functions.
- If they support multiple integration.
- Advantages and disadvantages of each (computational cost, when to use one over another, etc.).
This would be very useful for users.

**Comment by ev-br** [comment]
Pull requests most welcome!

**Comment by mdhaber** [comment]
There are plans to remove several of them, at which point the choice will be very easy. Right now, I can only recommend `quad` (and the higher dimensional variants).  

`quad_vec` is not as robust, but it supports vector values functions.

`qmc_quad` can be useful if the problem is too high-dimensional for quad to finish in reasonable time.

When I have time, I will propose removing `quadrature` and `romberg`. More information is here: https://github.com/scipy/scipy/issues/18574#issuecomment-1703320598.

**Comment by mdhaber** [comment]
If you can build from source, I'd suggest checking out the branch from https://github.com/scipy/scipy/pull/19173 and trying `_tanhsinh`, which might do what you're looking for. It is already merged - but with the shortcomings addressed by gh-19173 - so you can also try is with the latest nightly wheels. LMK what else you'd like to see when it becomes public.

**Comment by vnmabus** [comment]
I cannot use it right now, as my project is a library, and it should be able to work with several SciPy versions. That said, it looks promising. I do not think that support for multiple integration is done yet (is that planned?). Also the documentation regarding the dimensions of the arrays involved is a bit confusing right now.

**Comment by mdhaber** [comment]
> I do not think that support for multiple integration is done yet (is that planned?)

`nquad` just recursively calls `quad`, so we could do that.

> Also the documentation regarding the dimensions of the arrays involved is a bit confusing right now.

Which part? This?
> ``func`` must be an elementwise function: each element ``func(x)[i]`` must equal ``func(x[i])`` for all indices ``i``.

or the parts stating that input arrays need to be broadcastable?

**Comment by vnmabus** [comment]
> `nquad` just recursively calls `quad`, so we could do that.

That is a possibility for a first implementation. Having that as part of the unified interface also allows for more optimized implementations, if any.

> Which part? This?
> 
> > `func` must be an elementwise function: each element `func(x)[i]` must equal `func(x[i])` for all indices `i`.

This is clear right now, although it will need to be further clarified in the case of multiple integration. 

> or the parts stating that input arrays need to be broadcastable?

This part is what is more confusing for me, yeah.

**Comment by mdhaber** [comment]
> This part is what is more confusing for me, yeah.

Is the confusing part the word "broadcastable"? Does this help?
https://numpy.org/doc/stable/user/basics.broadcasting.html

**Comment by vnmabus** [comment]
No, I know what broadcasting is, I use that every day. But it is difficult to see the relationships between the dimensions of the different arrays involved (`input`, `output`, `a`, `b`, etc) as the dimensions are not explicit in the text. I would have preferred if you assigned names to the dimensions and added them to the documentation, e. g.:
- `f` accepts an array of dimension `n_points` $\times$ `dim_domain` and returns an array of dimension `n_points` $\times$ `dim_codomain`.
- `a` and `b` are 1D arrays of length `dim_domain` or floats (when the limits are the same for all variables).

**Comment by mdhaber** [comment]
There are no such requirements. It behaves almost as though one had applied `np.vectorize` to `quad`, which only works with scalars (except the output is a result object where each field is an array rather than the output being an array of result objects).

**Comment by vnmabus** [comment]
> LMK what else you'd like to see when it becomes public.

I tried quickly what happened just by copy-pasting code. Am I wrong in thinking that it does not allow vector-valued functions out of the box? I think that behavior should be changed, so it can be used as a replacement of `quad_vec`.

**Comment by mdhaber** [comment]
It depends.

Currently, the callable needs to written in a way that I call "strictly elementwise": "each element `func(x)[i]` must equal `func(x[i])` for all indices i." Actually, the requirement is a bit less restrictive than that - in addition to the variable `x`, you can also pass parameters. But when `x` and all the parameters are broadcasted to the same shape, the function needs to work elementwise. 

A vector-valued integrand like:

```python3
def f(x):
    return [sin(x), cos(x)]
```

does not satisfy the requirement. However, any vector-valued integrand can be expressed in a way that does satisfy the requirement. For example, the integrand could be written like:

```python3
def f(x, c1, c2):
    return c1*sin(x) + c2*cos(x)
```

and in addition to `x`, one would pass `c1 = np.asarray([1, 0])`, `c2 = np.asarray([0, 1])` as `args=(c1, c2)`. (There are more efficient ways, but this is conceptually simple.)

---

I realize this is cumbersome and inefficient for your use case, but our primary use case was univariate statistical distributions, so it was enough for our needs.

Actually, the original version of the framework this function uses (which was written for a scalar root-finding algorithm but has also been extended to scalar minimization, scalar differentiation, and scalar root bracketing) supported something like what you want. I changed that for efficiency (https://github.com/scipy/scipy/pull/18728#issuecomment-1612483600): if we can assume that the funciton works strictly elementwise, elements for which the integral estimate has already converged can be removed from further iterative refinement. At the time, I had offline discussions with the reviewer about how to offer both possibilities (efficiency vs callable flexibility), but we couldn't agree on an interface that would allow the best of both worlds, so we went with the simpler thing - make the function more efficient for the immediate use case.

Does this make sense? If so, would you be willing to suggest an interface that would allow the user to get the best of both worlds?

For instance, `_tanhsinh` could have a parameter `elementwise` which might be `False` by default, so it would natively support vector-valued, non-elementwise callables like you want. For a potential speed increase, users could specify `elementwise=True` when their function satisfies the elementwise requirement.

Alternatively, `_tanhsinh` could try to automatically detect whether the function works strictly elementwise, but there would be some overhead associated with that, and it might not be bulletproof.

Thanks for your thoughts!

**Comment by vnmabus** [comment]
I still do not get why that optimization is not applicable to vector valued functions (but I did not understand the finer details).

In my case I can make that `func(x[i]) == func(x)[i]`, but there are extra dimensions corresponding to the vector output (`func(x).shape == (len(x), dim_codomain)`), and this quadrature is not able to deal with that.

(I would also like if I could choose the dimension(s) that have to match with those of `x` through an `axis` parameter, but this is not strictly necessary as I can always move all extra dimensions to the end if necessary and back again).

**Comment by mdhaber** [comment]
I see that the description of the requirement is not complete. I will think about how to improve it.

> this quadrature is not able to deal with that.

It is currently able to solve any problem `quad_vec` can solve, but it might require a reformulation of the callable to express the problem.

> I still do not get why that optimization is not applicable to vector valued functions

One can get only part of the benefit if the function is vector valued because `_tanhsinh` has no way of stopping the callable from evaluating the components of the integrand vector for which the integral has already converged.

**Comment by mdhaber** [comment]
The close was a misclick on my cell phone. 

I'll think about this some more.
